File: WindowsRegular Expressions/4975sourcecode/chapter2/2-6/recipe2-6.vb
Imports System
Imports System.IO
Imports System.Text.RegularExpressions
Public Class Recipe

	Private Const hostRegex = "([a-z\d][-a-z\d]*[a-z\d]\.)*[a-z][-a-z\d]*[a-z]"
	Private Const portRegex = "(:\d{1,})?"
	Private Const pathRegex = "(/[^?<>#""\s]+)?"
	Private Const queryRegex = "(\?[^<>#""\s]+)?"

    Private Shared _Regex As Regex = New Regex("(?:(?<=^)|(?<=\s))((ht|f)tps?://" & hostRegex & portRegex & pathRegex & queryRegex & ")")

    Public Sub Run(ByVal fileName As String)
        Dim line As String
        Dim newLine As String
        Dim sr As StreamReader = File.OpenText(fileName)
        line = sr.ReadLine
        While Not line Is Nothing
            newLine = _Regex.Replace(line, "cow")
            Console.WriteLine("New string is: '{0}', original was: '{1}'", _
             newLine, _
             line)
            line = sr.ReadLine
        End While
        sr.Close()
    End Sub

    Public Shared Sub Main(ByVal args As String())
        Dim r As Recipe = New Recipe
        r.Run(args(0))
    End Sub

End Class

File: WindowsRegular Expressions/4975sourcecode/chapter2/2-6/recipe2-6.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

public class Recipe
{
	private static string hostRegex = @"([a-z\d][-a-z\d]*[a-z\d]\.)*[a-z][-a-z\d]*[a-z]";
	private static string portRegex = @"(:\d{1,})?";
	private static string pathRegex = @"(/[^\s?]+)?";
	private static string queryRegex = "(\\?[^<>#\"\\s]+)?";

	private static string fullRegex = @"(?:(?<=^)|(?<=\s))((ht|f)tps?://" + hostRegex + portRegex + pathRegex + queryRegex + ")" ;
	
	private static Regex _Regex = new Regex(fullRegex);
	
	public void Run(string fileName)
	{
		String line;
		String newLine;
		using (StreamReader sr = new StreamReader(fileName))
		{
			while(null != (line = sr.ReadLine()))
			{
				newLine = _Regex.Replace(line, "<a href=\"$1\">$1</a>");
				Console.WriteLine("New string is: '{0}', original was: '{1}'",
					newLine,
					line);
			}
		}
	}
	
	public static void Main( string[] args )
	{
		Recipe r = new Recipe();
		r.Run(args[0]);
	}
}

